polyply/tests/test_big_smile_parsing.py

```diff
@@ -0,0 +1,137 @@
+                        [(0, 1), (1, 2), (2, 3), (1, 4)]),
+                        # simple sequence two branches
+                        ("{[#PMA][#PMA][#PMA]([#PEO][#PEO])([#CH3])[#PMA]}",
+                        ["PMA", "PMA", "PMA", "PEO", "PEO", "CH3", "PMA"],
+                        [(0, 1), (1, 2), (2, 3), (3, 4), (2, 5), (2, 6)]),
+                        # simple linear sequence with expansion
+                        ("{[#PMA]|3}",
+                        ["PMA", "PMA", "PMA"],
+                        [(0, 1), (1, 2)]),
+                        # simple branch expension
+                        ("{[#PMA]([#PEO][#PEO][#OHter])|2}",
+                        ["PMA", "PEO", "PEO", "OHter",
+                         "PMA", "PEO", "PEO", "OHter"],
+                        [(0, 1), (1, 2), (2, 3),
+                         (0, 4), (4, 5), (5, 6), (6, 7)]
+                         ),
+                        # nested branched with expansion
+                        ("{[#PMA]([#PEO]|3)|2}",
+                        ["PMA", "PEO", "PEO", "PEO",
+                         "PMA", "PEO", "PEO", "PEO"],
+                        [(0, 1), (1, 2), (2, 3),
+                         (0, 4), (4, 5), (5, 6), (6, 7)]
+                         )
+))
+def test_res_pattern_to_meta_mol(smile, nodes, edges):
+    """
+    Test that the meta-molecule is correctly reproduced
+    from the simplified smile string syntax.
+    """
+    meta_mol = res_pattern_to_meta_mol(smile)
+    assert len(meta_mol.edges) == len(edges)
+    for edge in edges:
+        assert meta_mol.has_edge(*edge)
+    resnames = nx.get_node_attributes(meta_mol, 'resname')
+    assert nodes == list(resnames.values())
+
+@pytest.mark.parametrize('big_smile, smile, bonding',(
+                        # smiple symmetric bonding
+                        ("[$]COC[$]",
+                         "COC",
+                        {0: ["$"], 2: ["$"]}),
+                        # smiple symmetric bonding; multiple descript
+                        ("[$]COC[$][$1]",
+                         "COC",
+                        {0: ["$"], 2: ["$", "$1"]}),
+                        # named different bonding descriptors
+                        ("[$1]CCCC[$2]",
+                         "CCCC",
+                        {0: ["$1"], 3: ["$2"]}),
+                        # ring and bonding descriptors
+                        ("[$1]CC[$2]C1CCCCC1",
+                         "CCC1CCCCC1",
+                        {0: ["$1"], 1: ["$2"]}),
+                        # bonding descript. after branch
+                        ("C(COC[$1])[$2]CCC[$3]",
+                         "C(COC)CCC",
+                        {0: ["$2"], 3: ["$1"], 6: ["$3"]}),
+                        # left rigth bonding desciptors
+                        ("[>]COC[<]",
+                        "COC",
+                        {0: [">"], 2: ["<"]})
+))
+def test_tokenize_big_smile(big_smile, smile, bonding):
```

Comment:
How does this behave with specified atoms (rather than shorthand)?

`[>][NH2+]C`

Reply:
it works the same why now; test case was added. Hydrogens cannot have bonding descriptors unless they are themselves elementary hydrogen like [H].